aiida/common/progress_reporter.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+# pylint: disable=global-statement,unused-argument
+"""Provide a singleton progress reporter implementation.
+
+The interface is inspired by `tqdm <https://github.com/tqdm/tqdm>`,
+and indeed a valid implementation is::
+
+    from tqdm import tqdm
+    set_progress_reporter(tqdm, bar_format='{l_bar}{bar}{r_bar}')
+
+"""
+from contextlib import contextmanager
+from functools import partial
+from typing import Any, Callable, ContextManager, Iterator, Optional
+
+__all__ = ('get_progress_reporter', 'set_progress_reporter', 'progress_reporter_base', 'ProgressIncrementerBase')
+
+
+class ProgressIncrementerBase:
+    """A base class for incrementing a progress reporter."""
+
+    def set_description_str(self, text: Optional[str] = None, refresh: bool = True):
+        """Set the text shown by the progress reporter.
+
+        :param text: The text to show
+        :param refresh: Force refresh of the progress reporter
+
+        """
+
+    def update(self, n: int = 1):  # pylint: disable=invalid-name
+        """Update the progress counter.
+
+        :param n: Increment to add to the internal counter of iterations
+
+        """
+
+
+@contextmanager
+def progress_reporter_base(*,
```

Comment:
I haven't implemented my own context managers yet, so this might be a stupid question but:
 what is the purpose of these function arguments if they aren't passed to anything?

Reply:
Oh they are just not being used here because it is a dummy/null implementation, but in an actual implementation e.g. tqdm, they would do something a bit like:

```python
@contextmanager
def progress_reporter_base(total, desc):
   progress_bar = setup_progress_bar(total, desc)
   yield ProgressIterator(progress_bar)
   progress_bar.teardown()
```